### PR TITLE
docs/dev: Add backport instructions for contributors

### PR DESCRIPTION
**Pull request by cvybhu**
Add instructions on how to backport a feature to on older version of Scylla.

It contains a detailed step-by-step instruction so that people unfamiliar with intricacies of Scylla's repository organization can easily get the hang of it.

This is the guide I wish I had when I had to do my first backport.

I put it in backport.md because that looks like the file responsible for this sort of information.
For a moment I thought about `CONTRIBUTING.md`, but this is a really short file with general information, so it doesn't really fit there. Maybe in the future there will be some sort of unification (see #12126)

### CODE REVIEW

**Comment by cvybhu** [comment]
At first I put it in the wiki (https://github.com/scylladb/scylladb/wiki/Backporting), but I heard that it's better to put it in `docs/dev`, so I will remove the wiki page once the PR is merged.

File: docs/dev/backport.md [hunk 1]
```diff
@@ -36,3 +36,53 @@ keep the label
...
+Here are the instructions on how to backport some changes to an older version of Scylla.  
+It's the easiest to show it by an example, so let's look at PR [#12031](https://github.com/scylladb/scylladb/pull/12031), which fixed issue [#12014](https://github.com/scylladb/scylladb/issues/12014).
+
+### Wait for the original change to get merged
+First the original PR should get merged.
+In the example the author wanted to merge the branch `cvybhu:filter_multi_bug` into `scylladb:master`.
+Each PR is submitted on its own branch, which the PR author wants to merge into the `master` branch.
+Before the changes end up on `master`, they are first added to the `next` branch to run some additional tests. Later the `next` branch becomes `master` and the PR is officialy merged.
+
+### A backport PR
```
Comment (denesb) [inline docs/dev/backport.md:51]: The normal flow is the following:
* PR gets merged to next
* promotion merges next to master
* at this point, if the PR fixes an issue, said issue gets the "Backport Candidate" label
* at some later point a maintainer (Avi) goes through all issues that have said label and backport it to all affected releases, if they see it as a good backport candidate
* the "Backport Candidate" label is removed from the issue

A separate backport PR is required only if the backport is non-trivial: cherry-picking it onto the target branch generates non-trivial merge conflicts. In this case the original author of said PR is asked to prepare a backport PR.

File: docs/dev/backport.md [hunk 2]
```diff
@@ -36,3 +36,53 @@ keep the label
...
+In the example, the PR with the backport [#12086](https://github.com/scylladb/scylladb/pull/12086) wants to merge `cvybhu:filter_multi_bug_5.0` into `scylladb:branch-5.0`.
+Before the changes end up on `branch-5.0`, they are first added to the `next-5.0` branch to run some additional tests. Later the `next-5.0` branch becomes `branch-5.0` and the PR is officially backported.
+
+### Preparing the backport PR
+1. Go to the branch with the target backport version
+```bash
+git fetch origin branch-5.0
+git checkout -t origin/branch-5.0
+git submodule sync
+git submodule update --recursive
```
Comment (denesb) [inline docs/dev/backport.md:63]: You might need to do another `git pull origin branch-5.0` before this, as the first fetch might fail to get all the submodule changes (as it is executed before `git submodule sync`).

File: docs/dev/backport.md [hunk 3]
```diff
@@ -36,3 +36,53 @@ keep the label
...
+```
+2. Create a branch with the fix
+```bash
+git checkout -b my_fix_branch_5.0
+```
+
+3. Copy the changes to the branch, use the merge commit from `master`
+```bash
+git cherry-pick -m1 -x 2d2034ea28b8615a130146ad9780010d29d2fcc9
```
Comment (denesb) [inline docs/dev/backport.md:72]: Usually a better alternative is to cherry-pick individual commits. The backport will create a new merge commit.
This method will result in less conflicts at a time. On the other hand you might have to resolve similar conflicts multiple times.

Comment (denesb) [reply docs/dev/backport.md:72]: We should also add an example for a single-commit backport as well (basically the same, just without the `-m1` part).

**Review by denesb — CHANGES_REQUESTED** [review]

File: docs/dev/backport.md (hunk 3 above)
Comment (avikivity) [reply docs/dev/backport.md:72]: I usually try the merge-commit first, and only if the changes are too complicated I fall back to individual commits.

File: docs/dev/backport.md (hunk 1 above)
Comment (cvybhu) [reply docs/dev/backport.md:55]: Added a note that these instructions are for contributors, who were asked by maintainers to prepare a backport PR.

File: docs/dev/backport.md (hunk 2 above)
Comment (cvybhu) [reply docs/dev/backport.md:68]: Damn git is terrible.

Added `git pull` before `submodule update`

File: docs/dev/backport.md (hunk 3 above)
Comment (cvybhu) [reply docs/dev/backport.md:78]: Added a note that you can also use individual cherry-picks with the `-x` option.

File: docs/dev/backport.md (hunk 2 above)
Comment (denesb) [reply docs/dev/backport.md:68]: In the usual workflow you already have `branch-5.0` and you just check it out, then do the sync and update and pull.

**Review by denesb — APPROVED** [review]

File: docs/dev/backport.md (hunk 2 above)
Comment (cvybhu) [reply docs/dev/backport.md:68]: I'm no git expert, but my thinking was:
* use git fetch to fetch all commits from `branch-5.0` to the local repo
* use `git checkout -t origin/branch-5.0` to create a branch matching the remote one and checkout to it
* Now the local and version controlled submodule references could apparently be out of sync so I need to do `submodule sync`
* Then use `submodule update` to actually change the files in submodule directories to the right ones

I'm still not sure why the `git pull` is needed

**Comment by annastuchlik** [comment]
@cvybhu Thank you for this initiative to add the backport instructions - it's super useful!

File: docs/dev/backport.md (hunk 2 above)
Comment (avikivity) [reply docs/dev/backport.md:68]: I run `git submodule foreach git fetch --all` to get commits for all submodules for all repos for which I have remotes.

**Pull request closed without merging**